File: Java/Insertion_Sort.java
import java.util.*;

class Insertion_Sort
{
    //Creating method so that it will provide code reusability and maintability of code
    void sort(int arr[]) 
    { 
        int n = arr.length; //to take total number of element in array 
        for (int i = 1; i < n; ++i) { 
            int key = arr[i]; 
            int j = i - 1; 
            while (j >= 0 && arr[j] > key) { 
                arr[j + 1] = arr[j]; 
                j = j - 1; 
            } 
            arr[j + 1] = key; 
        } 
    } 
    //printing method
     static void printArray(int arr[]) 
    { 
        int length = arr.length; 
        for (int i = 0; i < length; ++i) 
            System.out.print(arr[i] + " "); 
  
        System.out.println(); 
    } 
    //Main method (Thread start from here  for execution )
    public static void main(String args[])
    {
        Scanner scanner=new Scanner(System.in);
        int arraySize=scanner.nextInt();
        System.out.println("Please! Enter Size of Array");
       int arr[] = new int[arraySize]; 
  
        Insertion_Sort object = new InsertionSort(); //here we are creating our object of our class although we can make our method static to access directly  without creating object
        object.sort(arr); 
         printArray(arr); 
}
/*
Problem: Sort the given array using insertion sort.

sample I/O:
Enter the size of array:6
Enter the elements:
34
67
23
14
65
22
Sorted array is:
14 22 23 34 65 67

Enter the size of array:5
Enter the elements:
98 87 76 65 54
Sorted array is:
54 65 76 87 98

Space Complexity: O(1)
Time Complxity: O(n^2) (worst case)S
*/

File: C-Plus-Plus/Insertion_sort.cpp
//Insertion sort

#include <bits/stdc++.h>

using namespace std;

//Insertion sort Function
void insertionSort(int arr[], int n)
{
    int num, j;
    for (int i = 1; i < n; i++)
    {
        j = i - 1;
        num = arr[i];
        while (j >= 0 && arr[j] > num)
        {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = num;
    }
}

//Function to print the array
void printArray(int arr[], int n)
{
    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;
}

int main()
{
    int n;

    cout << "Enter the number of elements: ";
    cin >> n;
    int arr[n];

    cout << "Enter the elements: ";
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
    }

    cout << "\nArray before sorting: ";
    printArray(arr, n);
    cout << "Array after sorting: ";
    insertionSort(arr, n);
    printArray(arr, n);

    return 0;
}

/*
INPUT:
Enter the number of elements: 5
Enter the elements: 5 6 0 -7 54
OUTPUT
----------------
Array before sorting: 5 6 0 -7 54 
Array after sorting: -7 0 5 6 54
----------------
Time complexity of Insertion sort : O(n^2)
Auxiliary Space complexity of Insertion sort : O(1)
*/
